python/cudf/cudf/utils/ioutils.py

```diff
@@ -2099,3 +2110,101 @@ def _read_byte_ranges(
 
     for worker in workers:
         worker.join()
+
+
+def _get_remote_bytes_all(
+    remote_paths, fs, *, blocksize=_BYTES_PER_THREAD_DEFAULT
+):
+    # TODO: Experiment with a heuristic to avoid the fs.sizes
+    # call when we are reading many files at once (the latency
+    # of collecting the file sizes is unnecessary in this case)
+    if max(sizes := fs.sizes(remote_paths)) <= blocksize:
+        # Don't bother breaking up individual files
+        return fs.cat_ranges(remote_paths, None, None)
+    else:
+        # Construct list of paths, starts, and ends
+        paths, starts, ends = map(
+            list,
+            zip(
+                *(
+                    (r, j, min(j + blocksize, s))
+                    for r, s in zip(remote_paths, sizes)
+                    for j in range(0, s, blocksize)
+                )
+            ),
+        )
+
+        # Collect the byte ranges
+        chunks = fs.cat_ranges(paths, starts, ends)
+
+        # Construct local byte buffers
+        # (Need to make sure path offsets are ordered correctly)
+        unique_count = dict(zip(*np.unique(paths, return_counts=True)))
+        offset = np.cumsum([0] + [unique_count[p] for p in remote_paths])
+        buffers = [
+            functools.reduce(operator.add, chunks[offset[i] : offset[i + 1]])
```

Comment:
*nit (non-blocking)*: I thought `reduce(add, foo)` is just `sum(foo)`, what am I missing?

Reply:
Yeah, this had me a bit confused as well. It turns out that `operator.add` will effectively join byte strings, but `sum` will require the intermediate values to be numeric values:

```python
import operator
assert operator.add(b"asdf", b"jkl;") == b'asdfjkl;'  # Assertion passes

assert sum([b"asdf", b"jkl;"]) == b'asdfjkl;'  # Raises
```
```
TypeError: unsupported operand type(s) for +: 'int' and 'bytes'
```